src/main/java/org/zeplinko/commons/lang/ext/core/AbstractResult.java

```diff
@@ -0,0 +1,129 @@
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+@SuppressWarnings("LombokGetterMayBeUsed")
+public abstract class AbstractResult<D, E> {
+    private final D data;
+
+    private final E error;
+
+    protected AbstractResult(D data, E error) {
+        this.data = data;
+        this.error = error;
+    }
+
+    /**
+     * Returns the data of the operation if it succeeded; otherwise, returns the
+     * null value.
+     *
+     * @return The success data, or null if the AbstractResult is a failure.
+     */
+    public D getData() {
+        return data;
+    }
+
+    /**
+     * Return the error object of the operation if it failed; otherwise, returns the
+     * null value.
+     *
+     * @return The error object, or null if the AbstractResult is successful.
+     */
+    public E getError() {
+        return error;
+    }
+
+    /**
+     * Checks if the AbstractResult represents success.
+     *
+     * @return {@code true} if the AbstractResult is successful, {@code false}
+     *         otherwise.
+     */
+    public boolean isSuccess() {
+        return !isFailure();
+    }
+
+    /**
+     * Checks if the AbstractResult represents failure.
+     *
+     * @return {@code true} if the AbstractResult is a failure, {@code false}
+     *         otherwise.
+     */
+    public abstract boolean isFailure();
+
+    /**
+     * Returns the data if the AbstractResult is successful; otherwise, returns the
+     * other specified value.
+     *
+     * @param other The value to return if the AbstractResult is a failure.
+     * @return The success data, or {@code other} if the AbstractResult is a
+     *         failure.
+     */
+    public D orElse(D other) {
+        return this.isFailure() ? other : this.getData();
+    }
+
+    /**
+     * Returns the data if the AbstractResult is successful; otherwise, applies the
+     * provided supplier to get the other value.
+     *
+     * @param otherSupplier The supplier function to provide the other value if the
+     *                      AbstractResult is a failure.
+     * @return The success data, or the value provided by {@code otherSupplier} if
+     *         the AbstractResult is a failure.
+     */
+    public D orElseGet(@Nonnull Supplier<? extends D> otherSupplier) {
+        Objects.requireNonNull(otherSupplier);
+        return this.isFailure() ? otherSupplier.get() : this.getData();
+    }
+
+    /**
+     * Returns the data if the AbstractResult is successful; otherwise, applies the
+     * provided function to the error and returns the AbstractResult.
+     *
+     * @param otherFunction The function to apply to the error if the AbstractResult
+     *                      is a failure.
+     * @return The success data, or the AbstractResult of applying
+     *         {@code otherFunction} to the error.
+     */
+    public D orElseGet(@Nonnull Function<? super E, ? extends D> otherFunction) {
```

Comment:
We should rename this differently. Here the behavior is not or else get, its actually recover from error or transfor / suppress the error and continue.

Reply:
Yeah we had discussion about this. We do have recover for that. But recover is chaining method and it will return Try/Result. orElseGet is a terminal method.

Motivation to keep the same name:
- Easy Discoverability.
- Method to cater specific requirements, whether you need to know the error to return specific value or return same value always.